### PR TITLE
Add support for decimal types to VectorFuzzer

### PR DESCRIPTION
**Pull request by karteekmurthys**
Add methods to generate random `UnscaledShortDecimal` and `UnscaledLongDecimal`
values. The values are capped within the range of precision provided.

### CODE REVIEW

File: velox/vector/fuzzer/VectorFuzzer.cpp [hunk 1]
```diff
@@ -703,12 +751,27 @@ TypePtr VectorFuzzer::randType(int maxDepth) {
...
+      LONG_DECIMAL(LongDecimalType::kMaxPrecision, 0)};
   static constexpr int kNumScalarTypes =
       sizeof(kScalarTypes) / sizeof(kScalarTypes[0]);
   // Should we generate a scalar type?
   if (maxDepth <= 1 || rand<bool>(rng_)) {
-    return kScalarTypes[rand<uint32_t>(rng_) % kNumScalarTypes];
+    auto type = kScalarTypes[rand<uint32_t>(rng_) % kNumScalarTypes];
+    if (isDecimalKind(type->kind())) {
+      // Generate precision in range [1, 18]
+      auto precision =
```
Comment (majetideepak) [inline velox/vector/fuzzer/VectorFuzzer.cpp:763]: this block can be further simplified
```
if (isDecimalKind(type->kind())) {
   // Generate precision in range [1, DecimalType max precision]
   auto precision = 1 + (rand<int8_t>(rng_) % (type->isShortDecimal() ? ShortDecimalType::kMaxPrecision : LongDecimalType::kMaxPrecision);
   
   // Generate scale in range [0, precision]
   auto scale = rand<int8_t>(rng_) % (precision + 1);
   return DECIMAL(precision, scale);
}
```

File: velox/vector/fuzzer/tests/VectorFuzzerTest.cpp [hunk 2]
```diff
@@ -96,6 +96,8 @@ TEST_F(VectorFuzzerTest, flatPrimitive) {
       DATE(),
       TIMESTAMP(),
       INTERVAL_DAY_TIME(),
+      LONG_DECIMAL(25, 3),
```
Comment (majetideepak) [inline velox/vector/fuzzer/tests/VectorFuzzerTest.cpp:99]: Do the precision and scale matter here?
Or will the above implementation randomize the precision and scale?

**Comment by karteekmurthys** [comment]
@mbasmanova @pedroerp 
Since fuzzer has support for decimals with this chage, the test `TYPED_TEST(UnsafeRowComplexDeserializerTests, Fuzzer)` tries to SerDe the LONG_DECIMAL type and fails. I was able to fix the serialization and deserialization part by extending the tempaltes, but the computing of offset expects the either the UnsafeRow fields to be of 8-bytes FixedWidth or of variable length. I would like to suggest to write 128-bit decimal as a variable length data instead of treating is as a fixed width type to minimize changes to computation of offset logic.

Do you have any other suggestions?

```
  /**
   * @param pos
   * @param isFixedWidth
   * @return the location of the fixed size data if isFixedWidth is true,
   * return the first unwritten word for variable length data otherwise.
   */
  char* getSerializationLocation(size_t pos, bool isFixedWidth) {
    setVariableLengthOffset(alignToFieldWidth(variableLengthOffset_));
    return isFixedWidth ? reinterpret_cast<char*>(&reinterpret_cast<uint64_t*>(
                              fixedLengthData_)[pos])
                        : buffer_ + variableLengthOffset_;
  }
```

**Comment by mbasmanova** [comment]
@karteekmurthys It sounds to me that UnsafeRowSerde doesn't support decimal types. Implementing this support may need to be a separate effort. For now, it might be sufficient to modify the test to not generate decimal types. This can be done by introducing a variation of VectorFuzzer::randRowType that takes a list of supported scalar types.

```
static TypePtr kSupportedTypes[]{
      BOOLEAN(),
      TINYINT(),
      SMALLINT(),
      INTEGER(),
      BIGINT(),
      REAL(),
      DOUBLE(),
      VARCHAR(),
      VARBINARY(),
      TIMESTAMP(),
      DATE(),
  };

const auto type = fuzzer.randRowType(kSupportedTypes);
```

**Comment by karteekmurthys** [comment]
@pedroerp @mbasmanova would you pls review this.

File: velox/row/tests/UnsafeRowDeserializerTest.cpp [hunk 3]
```diff
@@ -989,21 +989,34 @@ VectorFuzzer::Options fuzzerOptions() {
...
+      .useMicrosecondPrecisionTimestamp = true};
 }
 
 TYPED_TEST(UnsafeRowComplexDeserializerTests, Fuzzer) {
   if (std::is_same_v<TypeParam, UnsafeRowLegacyWrapper>) {
     LOG(WARNING) << "Disabled for legacy unsafe row deserializer";
     return;
   }
-  std::string buffer(100 << 20, '\0'); // Up to 100MB.
+  std::string buffer(200 << 20, '\0'); // Up to 200MB.
```
Comment (mbasmanova) [inline velox/row/tests/UnsafeRowDeserializerTest.cpp:1000]: Why is this change?

File: velox/vector/fuzzer/VectorFuzzer.h [hunk 4]
```diff
@@ -220,8 +220,12 @@ class VectorFuzzer {
   //
   // There are no options to control type generation yet; these may be added in
   // the future.
-  TypePtr randType(int maxDepth = 5);
-  RowTypePtr randRowType(int maxDepth = 5);
+  TypePtr randType(
+      int maxDepth = 5,
+      const std::vector<TypePtr>& scalarTypes = {});
```
Comment (mbasmanova) [inline velox/vector/fuzzer/VectorFuzzer.h:225]: Would you document the new parameter?

File: velox/vector/fuzzer/VectorFuzzer.cpp [hunk 5]
```diff
@@ -456,8 +495,14 @@ VectorPtr VectorFuzzer::fuzzFlatPrimitive(
     // First, fill it with random values.
     // TODO: We should bias towards edge cases (min, max, Nan, etc).
     auto kind = vector->typeKind();
-    VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
-        fuzzFlatPrimitiveImpl, kind, vector, rng_, opts_);
+    if (kind == TypeKind::SHORT_DECIMAL) {
```
Comment (mbasmanova) [inline velox/vector/fuzzer/VectorFuzzer.cpp:498]: Why do we need these special cases? Why can't we simply provide template specializations for fuzzFlatPrimitiveImpl?

File: velox/vector/fuzzer/VectorFuzzer.cpp [hunk 6]
```diff
@@ -686,12 +731,29 @@ BufferPtr VectorFuzzer::fuzzNulls(vector_size_t size) {
 
 variant VectorFuzzer::randVariant(const TypePtr& arg) {
   VELOX_CHECK(arg->isPrimitiveType());
+  if (isDecimalKind(arg->kind())) {
```
Comment (mbasmanova) [inline velox/vector/fuzzer/VectorFuzzer.cpp:734]: I'm removing randVariant method in #3834

File: velox/vector/fuzzer/VectorFuzzer.cpp [hunk 7]
```diff
@@ -686,12 +731,29 @@ BufferPtr VectorFuzzer::fuzzNulls(vector_size_t size) {
...
-TypePtr VectorFuzzer::randType(int maxDepth) {
-  static TypePtr kScalarTypes[]{
+std::pair<int8_t, int8_t> VectorFuzzer::randPrecisionScale(TypeKind kind) {
+  VELOX_DCHECK(isDecimalKind(kind));
+  // Generate precision in range [1, Decimal type max precision]
+  auto precision = 1 +
+      rand<int8_t>(rng_) %
+          (kind == TypeKind::SHORT_DECIMAL ? ShortDecimalType::kMaxPrecision
+                                           : LongDecimalType::kMaxPrecision);
+  // Generate scale in range [0, precision]
```
Comment (mbasmanova) [inline velox/vector/fuzzer/VectorFuzzer.cpp:748]: typos?

File: velox/vector/fuzzer/VectorFuzzer.cpp [hunk 8]
```diff
@@ -703,12 +765,20 @@ TypePtr VectorFuzzer::randType(int maxDepth) {
       VARBINARY(),
       TIMESTAMP(),
       DATE(),
-  };
-  static constexpr int kNumScalarTypes =
-      sizeof(kScalarTypes) / sizeof(kScalarTypes[0]);
+      SHORT_DECIMAL(ShortDecimalType::kMaxPrecision, 0),
```
Comment (mbasmanova) [inline velox/vector/fuzzer/VectorFuzzer.cpp:768]: Can we generate decimals with different precisions and scales?

**Review by mbasmanova — COMMENTED** [review]
@karteekmurthys Karteek, I took a quick look and have some high-level comments.

Let's extract changes to randType, randRowType and UnsafeRowTest into a separate PR to reduce the scope of the changes and make them easier to review.

The Fuzzer seems to be generating only decimal types with max precision. Can we generate types with different precisions and scales?

File: velox/vector/fuzzer/VectorFuzzer.cpp (hunk 8 above)
Comment (karteekmurthys) [reply velox/vector/fuzzer/VectorFuzzer.cpp:763]: This is just a place holder. The precision and scale will be randomly generated here: https://github.com/facebookincubator/velox/pull/3768/files/34d539eca6b0616c69ce00d0a2e5c6aaf93ac265#diff-68474a90a74f69bf4c82ba43c95a744abcf7e95c1ff6ffd31cae8cda84b5250bR778

Explained in comments.

File: velox/row/tests/UnsafeRowDeserializerTest.cpp (hunk 3 above)
Comment (karteekmurthys) [reply velox/row/tests/UnsafeRowDeserializerTest.cpp]: Undone

File: velox/vector/fuzzer/VectorFuzzer.cpp [hunk 9]
```diff
@@ -703,12 +751,27 @@ TypePtr VectorFuzzer::randType(int maxDepth) {
       VARBINARY(),
       TIMESTAMP(),
       DATE(),
-  };
+      SHORT_DECIMAL(ShortDecimalType::kMaxPrecision, 0),
+      LONG_DECIMAL(LongDecimalType::kMaxPrecision, 0)};
```
[the remark this answers is not in the document]
Comment (karteekmurthys) [reply velox/vector/fuzzer/VectorFuzzer.cpp]: The format-fix does this. I will try changing it.

File: velox/vector/fuzzer/VectorFuzzer.cpp (hunk 6 above)
Comment (karteekmurthys) [reply velox/vector/fuzzer/VectorFuzzer.cpp]: Ack.

File: velox/vector/fuzzer/VectorFuzzer.cpp (hunk 5 above)
Comment (karteekmurthys) [reply velox/vector/fuzzer/VectorFuzzer.cpp]: There are some specializations not yet implemented for decimals and leads to errors like this:

```
/Users/karteek/workspace/velox/./velox/type/Variant.h:750:47: note: in instantiation of function template specialization 'facebook::velox::VariantConverter::convert<facebook::velox::TypeKind::SHORT_DECIMAL>' requested here
    return VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(convert, toKind, value);
                                              ^
/Users/karteek/workspace/velox/./velox/type/Variant.h:279:3: note: candidate constructor not viable: no known conversion from 'typename TypeTraits<(TypeKind)'\f'>::NativeType' (aka 'facebook::velox::UnscaledShortDecimal') to 'const typename detail::VariantTypeTraits<TypeKind::BOOLEAN>::native_type' (aka 'const bool') for 1st argument
  VELOX_VARIANT_SCALAR_MEMBERS(TypeKind::BOOLEAN)
  ^
/Users/karteek/workspace/velox/./velox/type/Variant.h:203:18: note: expanded from macro 'VELOX_VARIANT_SCALAR_MEMBERS'
  /* implicit */ variant(                                             \
                 ^
/Users/karteek/workspace/velox/./velox/type/Variant.h:279:3: note: candidate constructor not viable: no known conversion from 'typename TypeTraits<(TypeKind)'\f'>::NativeType' (aka 'facebook::velox::UnscaledShortDecimal') to 'typename detail::VariantTypeTraits<TypeKind::BOOLEAN>::native_type' (aka 'bool') for 1st argument
  VELOX_VARIANT_SCALAR_MEMBERS(TypeKind::BOOLEAN)
  ^
/Users/karteek/workspace/velox/./velox/type/Variant.h:207:18: note: expanded from macro 'VELOX_VARIANT_SCALAR_MEMBERS'
  /* implicit */ variant(                                             \
                 ^
```

Using `VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH_ALL` seems to do the job. There aren't too many references to this macro.

File: velox/vector/fuzzer/VectorFuzzer.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (karteekmurthys) [reply velox/vector/fuzzer/VectorFuzzer.cpp]: The format check fails if `}` is in newline:

```
+      .useMicrosecondPrecisionTimestamp = true};
 }
 
 TYPED_TEST(UnsafeRowComplexDeserializerTests, Fuzzer) {
diff --git a/velox/vector/fuzzer/VectorFuzzer.cpp b/velox/vector/fuzzer/VectorFuzzer.cpp
index 13d0da43..fbe827b8 100644
--- a/velox/vector/fuzzer/VectorFuzzer.cpp
+++ b/velox/vector/fuzzer/VectorFuzzer.cpp
@@ -761,8 +761,7 @@ TypePtr VectorFuzzer::randType(
       // Precision and scale are just placeholders and will be
       // replaced with random values.
       SHORT_DECIMAL(ShortDecimalType::kMaxPrecision, 0),
-      LONG_DECIMAL(LongDecimalType::kMaxPrecision, 0)
-  };
+  
```

File: velox/vector/fuzzer/VectorFuzzer.cpp (hunk 7 above)
Comment (karteekmurthys) [reply velox/vector/fuzzer/VectorFuzzer.cpp:741]: Sorry, not able to spot the typo here.
The scale will in the range 0 to precision inclusive.

**Comment by karteekmurthys** [comment]
@mbasmanova @majetideepak would you please review this.

File: velox/vector/fuzzer/VectorFuzzer.h [hunk 10]
```diff
@@ -222,10 +222,16 @@ class VectorFuzzer {
   // limits the maximum level of nesting for complex types. maxDepth <= 1 means
   // no complex types are allowed.
   //
-  // There are no options to control type generation yet; these may be added in
-  // the future.
-  TypePtr randType(int maxDepth = 5);
-  RowTypePtr randRowType(int maxDepth = 5);
+  // @param int maxDepth Depth of a complex type.
+  // @param const std::vector<TypePtr> scalarTypes List of scalar types to be
```
Comment (mbasmanova) [inline velox/vector/fuzzer/VectorFuzzer.h:226]: Drop 'const std::vector<TypePtr>'.

File: velox/vector/fuzzer/VectorFuzzer.h [hunk 11]
```diff
@@ -254,6 +260,11 @@ class VectorFuzzer {
       RowVectorPtr rowVector,
       const std::vector<column_index_t>& columnsToWrapInLazy);
 
+  /// Generates random precision in range [1, max precision for decimal
+  /// TypeKind] and scale in range [0, random precision generated].
+  /// @param kind must be a decimal type kind.
+  std::pair<int8_t, int8_t> randPrecisionScale(TypeKind kind);
```
Comment (mbasmanova) [inline velox/vector/fuzzer/VectorFuzzer.h:266]: Should this be a private method?

File: velox/vector/fuzzer/VectorFuzzer.cpp [hunk 12]
```diff
@@ -715,12 +763,22 @@ TypePtr VectorFuzzer::randType(int maxDepth) {
...
+      // replaced with random values.
+      SHORT_DECIMAL(ShortDecimalType::kMaxPrecision, 0),
+      LONG_DECIMAL(LongDecimalType::kMaxPrecision, 0)};
+  auto scalarTypes =
+      supportedScalarTypes.size() == 0 ? kScalarTypes : supportedScalarTypes;
+  const int numScalarTypes = sizeof(scalarTypes) / sizeof(scalarTypes[0]);
   // Should we generate a scalar type?
   if (maxDepth <= 1 || rand<bool>(rng_)) {
-    return kScalarTypes[rand<uint32_t>(rng_) % kNumScalarTypes];
+    auto type = scalarTypes[rand<uint32_t>(rng_) % numScalarTypes];
```
Comment (mbasmanova) [inline velox/vector/fuzzer/VectorFuzzer.cpp:775]: This logic is confusing to me. I wonder if a cleaner design would be to have a list of scalar type kinds and have a method to generate a type for a given type kind. The supportedScalarTypes parameter would also be a list of type kinds, not types.

Comment (karteekmurthys) [reply velox/vector/fuzzer/VectorFuzzer.cpp:775]: That sounds better what we have now. I will add that change.

File: velox/vector/fuzzer/VectorFuzzer.h (hunk 11 above)
Comment (pedroerp) [reply velox/vector/fuzzer/VectorFuzzer.h:276]: +1

**Review by pedroerp — COMMENTED** [review]
Just went to the code and it looks good to me. Please check double check you addressed Masha's comments, and thanks for moving the randType() discussion to a separate Issue/PR!

File: velox/vector/fuzzer/VectorFuzzer.h (hunk 11 above)
Comment (karteekmurthys) [reply velox/vector/fuzzer/VectorFuzzer.h:276]: Done.

**Comment by karteekmurthys** [comment]
@mbasmanova would you please review this.

File: velox/vector/fuzzer/tests/VectorFuzzerTest.cpp [hunk 13]
```diff
@@ -226,6 +227,18 @@ TEST_F(VectorFuzzerTest, constants) {
   ASSERT_EQ(VectorEncoding::Simple::CONSTANT, vector->encoding());
   ASSERT_FALSE(vector->mayHaveNulls());
 
+  auto shortDecimalType = fuzzer.randShortDecimalType();
```
Comment (mbasmanova) [inline velox/vector/fuzzer/tests/VectorFuzzerTest.cpp:230]: How do you plan to use these new methods? Don't we need this logic integrated into randType()?

Comment (karteekmurthys) [reply velox/vector/fuzzer/tests/VectorFuzzerTest.cpp:230]: It needs to be and I have opened an issue for the same. Also, you suggested to isolate the changes to randType in a separate. I have opened an issue for this: https://github.com/facebookincubator/velox/issues/3942.

This change only adds capability to generate random DECIMALS and random unscaled values.
The random unscaled values are getting use in this PR:https://github.com/facebookincubator/velox/pull/3755/files now.

**Pull request closed without merging**